### ISSUE TITLE
Regression caused by #1594

### ISSUE THREAD

**Issue by mbechard** [KhronosGroup/MoltenVK#1672]
Since #1594 was merged, specifically 745af5379, I'm seeing a metal assert in a very specific case. When trying to use vkCmdDrawIndexedIndirect, I get the assert
`failed assertion 'Set Vertex Buffer Offset Validation index(0) must have an existing buffer.'`

Tracing through the MVK code, I can see there is a setVertexBuffer: call that is correctly setting the index(0) buffer, and doing a regular draw call. Then later on with the same render command encoder, it tries to do setVertexBufferOffset: when setting up for an indirect draw, which throws that assert. There are pipeline switches that have occurred since the first draw call and the indirect draw call. Are there cases where binding a new pipeline will invalidate the vertex binding, and that's not being handled by this PR?
What could cause the vertex binding to get lost like this? Is there any Metal operations that could invalidate the index?
Any tips on how I could help debug this more? I've set a breakpoint in MVKResourcesCommandEncoderState::encodeBinding to try to catch any there calls that may be overwriting the binding, but not seeing anything questionable.

One thing that is happening between the first setVertexBuffer: to the render encoder, and the setVertexBufferOffset:, is the the same buffer is getting bound to a compute encoder. I would think that should be fine though. The computer encoder will get committed before the render encoder does, but this assert is happening during the recording of the commands anyways, not the execution.

Attn: @js6i 

Thanks

**Comment by mbechard** [comment]
Adding to this, it doesn't seem limited to the binding becoming invalidated (unbound). I have another case where the binding size/offset isn't what MoltenVK think it is, so a validation error comes up there as well (not enough data bound for the shader requirement). 
So far no luck finding where the core issue is, but just always treating the bindings as dirty avoids the issue, so I do think it's related to this optimization.

**Comment by billhollings** [comment]
> I have another case...

In this second case, is there also a flip to a compute encoder between two render encoders? That is...is this behavior consistent with the original problem above?

**Comment by mbechard** [comment]
I just tried commenting out all calls to vkCmdDispatch(), and the issue still occurs (the second case). So I think it's not specifically related to mixing those.

**Comment by billhollings** [comment]
@mbechard 

I'm curious. How did you determine that 745af53 is the problematic commit?

And...

> just always treating the bindings as dirty avoids the issue

What did you change in the MoltenVK code to do this...to treat the bindings as dirty?

**Comment by aglab2** [comment]
I am having issues with PR 1594 too. Metal assertion occurs in parallel-n64 for macOS:
`Compute Function(main0): argument _5227[0] from buffer(0) with offset(0) and length(16) has space for 16 bytes, but argument has a length(32).`

I have pinpointed it to https://github.com/KhronosGroup/MoltenVK/blob/master/MoltenVK/MoltenVK/Commands/MVKMTLResourceBindings.h#L77-L91, replacing the condition to line 78 to `true` fixes the assert but I am not sure why exactly it helps.

**Comment by billhollings** [comment]
Do these render passes involve tessellation?

**Comment by aglab2** [comment]
I very much doubt, it is just an N64 emulator. The source code is at https://gitlab.com/aglab2/parallel-n64/-/tree/moltenvk_noupstream, unfortunately I am not very knowledged in Vulkan to understand it entirely. I also noticed that just adding a check for `other.isDirty` is enough to fix the issue but I think it defeats the purpose of the optimization so I doubt it is the solution.
```
    inline void update(const MVKMTLBufferBinding &other) {
        if (other.isDirty) {
          isDirty = true;
        }
        if (mtlBuffer != other.mtlBuffer || size != other.size || isInline != other.isInline) {
            mtlBuffer = other.mtlBuffer;
            size = other.size;
            isInline = other.isInline;
            offset = other.offset;

            justOffset = false;
            isDirty = true;
        } else if (offset != other.offset) {
            offset = other.offset;
            justOffset = !isDirty || justOffset;
            isDirty = true;
        }
    }
```

**Comment by mbechard** [comment]
I determined it was the problematic commit by recompiling at different points until things started to break. The source at 745af5379c69 will generate the metal assertion, while the source at the previous commit 334e0cee69c60c will not.
I can also avoid the assertion if I ensure that MVKMTLBufferBinding::update() always takes the first if clause, which I assume pretty much neutralizes this commit.
No Tess shaders in my system either.

**Comment by billhollings** [comment]
@mbechard @aglab2 

PR #1677 possibly fixes this. I can't be sure, because I haven't been able to test your situations.

That PR is a fix to when a push constant in one pipeline used the same Metal buffer index as a descriptor in a previous pipeline, and if the descriptor was used again in a later pipeline, and only the offset has changed, the descriptor would use the `MTLBuffer` from the push constant.

Please retest using that PR, and close this issue if it fixes the problem you encountered.

**Comment by aglab2** [comment]
Did not fix it for me, still the same assert. If you want to replicate my setup, you may use https://parallel-launcher.ca/ with "Parallel" selected as a graphics plugin and launch any ROM. libMoltenVK.dylib is placed in `~/Library/parallel-launcher/data/retroarch/RetroArch.app/Contents/Frameworks/libMoltenVK.dylib`, app is just adhoc signed. To test it, you may launch the app in a debugger:
```
METAL_DEVICE_WRAPPER_TYPE=1 lldb ~/Library/parallel-launcher/data/retroarch/RetroArch.app/Contents/MacOS/RetroArch
(lldb) r -L ~/Library/parallel-launcher/data/retro-data/cores/parallel_n64_next_libretro.dylib --config ~/Library/parallel-launcher/data/retroarch.cfg "~/Downloads/test.z64"
...
validateComputeFunctionArguments:771: failed assertion `Compute Function(main0): argument _5227[0] from buffer(0) with offset(0) and length(16) has space for 16 bytes, but argument has a length(32).'
```

**Comment by billhollings** [comment]
@aglab2 

Thanks for providing the environment. 

Where do I found the `text.z64` game file? Or other game files?

**Comment by aglab2** [comment]
Here is the test ROM you may use: [test.z64.zip](https://github.com/KhronosGroup/MoltenVK/files/9312973/test.z64.zip). It should look like this:

![Screenshot 2022-08-12 at 8 48 39 AM](https://user-images.githubusercontent.com/6104625/184265353-984cc800-97f1-4de9-bde3-a803436d1c55.png)

**Comment by js6i** [comment]
One more thing to check, I think the second line:
```
inline void update(const MVKMTLBufferBinding &other) {
    if (mtlBuffer != other.mtlBuffer || size != other.size || isInline != other.isInline) {
```
should instead read
```
    if (mtlBuffer != other.mtlBuffer || size != other.size || isInline || other.isInline) {
```
it doesn't matter if inline bindings have the same pointer and size, their data may be different.

**Comment by billhollings** [comment]
> it doesn't matter if inline bindings have the same pointer and size, their data may be different.

Good point. And `isInline != other.isInline` is effectively the same test as `mtlBuffer != other.mtlBuffer`, since inline and non-inline would necessarily have different `mtlBuffer` values.

As would `isInline || other.isInline` if the values of those two booleans are different. So perhaps the only thing that matters in the inline test is that we are now inline (`other.isInline`):

    if (mtlBuffer != other.mtlBuffer || size != other.size || other.isInline) {

**Comment by billhollings** [comment]
@mbechard @aglab2

Okay. This should be fixed now in the updated #1677. Verified with `test.z64.zip`.

In the previous version of that PR, for the descriptor binding that was being overridden, I was marking it as dirty too early. Fixed now.

The PR also includes [my version](https://github.com/KhronosGroup/MoltenVK/issues/1672#issuecomment-1212963873) of @js6i 's [insight](https://github.com/KhronosGroup/MoltenVK/issues/1672#issuecomment-1212933456) about tracking inline buffers.

**Comment by mbechard** [comment]
Confirmed this fixes the issue for me, thanks!
